ophyd/log.py

```diff
@@ -150,44 +203,70 @@ def set_handler(file=sys.stdout, datefmt='%H:%M:%S', color=True):
         Date format. Default is ``'%H:%M:%S'``.
     color : boolean
         Use ANSI color codes. True by default.
-
+    level : str or int
+        Python logging level, given as string or corresponding integer.
+        Default is 'INFO'.
+    backupCount : int
+        Number of historical log files to keep. Default is 4.
     Returns
     -------
     handler : logging.Handler
         The handler, which has already been added to the 'ophyd' logger.
-
     Examples
     --------
     Log to a file.
-
-    >>> set_handler(file='/tmp/what_is_happening.txt')
-
+    >>> config_ophyd_logging(file='/tmp/what_is_happening.txt')
     Include the date along with the time. (The log messages will always include
     microseconds, which are configured separately, not as part of 'datefmt'.)
-
-    >>> set_handler(datefmt="%Y-%m-%d %H:%M:%S")
-
+    >>> config_ophyd_logging(datefmt="%Y-%m-%d %H:%M:%S")
     Turn off ANSI color codes.
-
-    >>> set_handler(color=False)
+    >>> config_ophyd_logging(color=False)
+    Increase verbosity: show level DEBUG or higher.
+    >>> config_ophyd_logging(level='DEBUG')
     """
     global current_handler
     if isinstance(file, str):
-        handler = logging.FileHandler(file)
+        handler = logging.handlers.TimedRotatingFileHandler(
+            filename=file,
+            when="W0",  # roll over every Monday
+            backupCount=backupCount,  # keep the most recent "backupCount" log files
+        )
     else:
         handler = logging.StreamHandler(file)
+    levelno = validate_level(level)
+    handler.setLevel(levelno)
     if color:
-        format = color_log_format
+        log_format = color_log_format
     else:
-        format = plain_log_format
-    handler.setFormatter(
-        LogFormatter(format, datefmt=datefmt))
+        log_format = plain_log_format
+    handler.setFormatter(LogFormatter(log_format, datefmt=datefmt))
+
     if current_handler in logger.handlers:
         logger.removeHandler(current_handler)
     logger.addHandler(handler)
+
+    if current_handler in control_layer_logger.handlers:
+        control_layer_logger.removeHandler(current_handler)
+    control_layer_logger.addHandler(handler)
     current_handler = handler
+
+    if logger.getEffectiveLevel() > levelno:
+        logger.setLevel(levelno)
+
+    if control_layer_logger.getEffectiveLevel() > levelno:
+        control_layer_logger.setLevel(levelno)
+
     return handler
 
 
 # Add a handler with the default parameters at import time.
-current_handler = set_handler()
+config_ophyd_logging()
```

Comment:
We used to think it was convenient to set up a default handler at import time, but we have since learned differently. The Python logging documentation advises strongly against libraries adding handlers at import time. In practice it seems tog get in the way as often as it helps; it's better to just provide and advertise a convenience function for users to call to add special logging, and then set it up to suit at the application level.

For example, at NSLS-II, we plan to call set up handlers in the profiles, ensuring that log messages are always written to some standard location.

Reply:
(Note: this section was copied out of caproto, which used to set up a handler at import time but now correctly leaves it to the application to do.)

Reply:
Two (at least) trains of thought on where to write the logs.

1. well-known location
1. user's working directory

Guiding the choice is the answer to *Who needs this logging?*  Initially, the developer who is expects a predictable location.  Longer term, the user needs a record of what they have typed and in what sequence.  Writing the logs to the user directory means they can copy that directory (which may contain other content such as *.py files with their custom plans).

If the developer can tolerate logs going into the user directory, that should be the default.  **But**, the location should be configurable.

Another configurable should be 

* un/limited number of logs
* un/limited length of one log file

Reply:
Agreed. Bluesky will not make any choice by default, leaving it up to the application (or the user's IPython profile) but should probably provide some recommendations. At a recent group meeting, we discussed that `$XDG_DATA_HOME/bluesky/logs` (i.e. `~/.local/share/bluesky/logs`) is probably the "proper" place. Unfortunately, since we at NSLS-II have very slow shared home directory performance, we will probably use a world-writable system location, maybe under `/var/log`, but only until our storage situation is fixed. As for limited log file length and number, we will probably rely on and recommend https://docs.python.org/3/library/logging.handlers.html#logging.handlers.TimedRotatingFileHandler.

Reply:
Another possibility is configuring _two_ handlers, one for the user's local directory and one into a system used by the facility support team, which might be a system location or a network-accessible log capture service like Elastic or Graylog.

Reply:
I agree writing the logs to the user's home directory is very reasonable.

The question of further configuration such as number of logs and length of one log files runs into another thing I have been thinking about which is combined configuration of bluesky and ophyd logging (and maybe others). Maybe we should supply a handler to the `config_ophyd_logging` method? Or have a `add_ophyd_logging_handler`?

Reply:
At that point, I think users should just use the standard Python logging API. Our config_X_logging methods are just shortcuts to help non-developers turn on some extra logging to the terminal or a file as a one-liner without knowing the details. For facility configurations, we can assume more background knowledge.

Reply:
This discussion motivates me to make changes in [stdlogpj](https://github.com/prjemian/stdlogpj). 

> we discussed that `$XDG_DATA_HOME/bluesky/logs` (i.e. `~/.local/share/bluesky/logs`) is probably the "proper" place.  

This sounds like a facility-dependency, not general for any facility.

Reply:
> This sounds like a facility-dependency, not general for any facility.

Yes. I believe this discussion has no bearing on what should go in the _library_ code. Facilities can make their own decisions and implement them using `import logging; logging.getLogger('ophyd').addHandler(...)` without any special code in ophyd itself. At most, we could decide to document some recommendations or examples to help folks navigate the Python logging framework, [as we have done in the caproto docs](https://caproto.github.io/caproto/loggers.html).